rules/do-you-optimize-tinacms-project/rule.md

```diff
@@ -0,0 +1,106 @@
+---
+type: rule
+title: Do you optimize your TinaCMS project for clarity, performance, and reliable builds?
+seoDescription: Learn effective strategies to structure and optimize your
+  TinaCMS project for clarity, performance, and successful builds.
+uri: do-you-optimize-tinacms-project
+authors:
+  - title: "Gilles Pothieu"
+    url: https://www.ssw.com.au/people/gilles-pothieu/
+created: 2024-08-28T16:15:00.000Z
+guid: 3c59f56f-34b0-47a7-902a-718eb340d2e5
+---
+Structuring and optimizing your TinaCMS project is essential to achieve clarity, enhance performance, and prevent build failures. Poorly optimized projects can lead to slow site performance, increased server load, and even failed builds due to excessive or inefficient data requests.
+
+Let’s explore how to structure your project effectively and apply best practices to boost performance both in runtime and during the build process.
+
+<!--endintro-->
+
+## 1. Structuring your TinaCMS Architecture
+
+When working with large datasets or generating multiple subcomponents, following best practices is crucial to maintain performance and clarity.
+
+### ❌ Bad practices
+
+* **Using deeply nested schemas with nested references**
+
+  * Complex and deeply nested schemas increase the complexity of the project, making it harder to manage and more prone to build failures
+  * They can also lead to inefficient data fetching, further slowing down both runtime and build processes
+
+### ✅ Good practices
+
+* **Making a single request at a top-level server component and using React Context or a state management library**
+
+  * Data fetched at the top level can be stored in a React Context or a global state management solution (e.g., [Redux](https://redux.js.org/)). This allows all components to access the data without the need to pass props manually
+  * This approach ensures better scalability, as subcomponents can access the necessary data directly from the context or store, eliminating redundant API calls and avoiding prop drilling
+
+::: good
+```js
+export default async function Home({ params }: HomePageProps) {
+  const location = params.location;
+  
+  // Appel à l'API ou client GraphQL
+  const websiteProps = await client.queries.website({
+    relativePath: `${location}/website.md`,
+  });
+
+  // Imaginons que websiteProps contient ces données
+  const { conferencesData, footerData, speakers } = websiteProps.data;
+
+  return (
+      <ConferenceContext.Provider value={conferencesData}>
+        <FooterContext.Provider value={footerData}>
+          <PageTransition>
+            <HomeComponent speakers={speakers}/>
+          </PageTransition>
+        </FooterContext.Provider>
+      </ConferenceContext.Provider>
+  );
+}
+
+// Génération des paramètres statiques pour SSG
```

Comment:
Please change this to English, or remove the comment

Reply:
✅ Done